docs/manpages/tracker-export.1
.TH tracker-export 1 "Mar 2020" GNU "User Commands"

.SH NAME
tracker-export \- Export all data from a Tracker database.

.SH SYNOPSIS
\fBtracker export\fR [\fIoptions\fR...]

.SH DESCRIPTION
.B tracker export
exports all data stored in a Tracker database, in Turtle format.

The output is intended to be machine-readable, not human readable.
Use a tool such as rapper(1) to convert the data to different formats.

.SH OPTIONS
.TP
.B \-g, \-\-show-graphs
This flag causes the relevant GRAPH statements to be output along with
the data.

In this mode the output is TriG syntax rather than Turtle, due to
the extra GRAPH statements. Some tools which understand Turtle do not
understand TriG.

.SH EXAMPLES
.TP
Export all data from Tracker Index and prettify the output using rapper(1).

.BR
.nf
$ tracker export -b org.freedesktop.Tracker1.Miner.Files | rapper - -I . -i turtle -o turtle
.fi

.SH SEE ALSO
.BR tracker-import (1),
.BR tracker-sparql (1).


NEWS
NEW in 2.99.1 - 2020-05-03
==========================

  Highlighted updates compared to 2.x:
  ------------------------------------
  * A full implementation of SPARQL 1.1
  * A change of paradigm. There is no central store, services are
    free to create their own, and to make those an endpoint for public
    access. Applications may syndicate queries across external endpoints
    through the SERVICE syntax.
  * No tracker-store
  * No libtracker-miner
  * No libtracker-control
  * API changes
    - Explicit API to create a local store, or proxy a DBus/remote one
    - TrackerEndpoint proxies local stores out of process
    - separate python library to write tests
  * A base and reference ontologies of our own
    - Officially it's no longer Nepomuk
    - No more broken links
  * An extensible CLI tool, with a generic set of core subcommands
  * Parallel installable to tracker 2.x
  * Documentation
  * Tests
  * Many many others

NEW in 2.3.0 - 2019-09-09
=========================

  Highlighted changes since 2.2.x:
  --------------------------------
  * Support for storing external references for other services
    (eg. musicbrainz)

NEW in 2.2.99.1 - 2019-09-05
============================

  * Handle circular references in TrackerResource
  * Removed stress tests
  * Handle application/x-zero-size
  * Refactors and cleanups in functional testing infrastructure
  * Documentation improvements
  * Make tracker_sparql_escape_string() escape single quotes
  * Don't make tracker:referenceSource a subproperty of nie:identifier

Translations: eu

NEW in 2.2.99.0 - 2019-08-07
============================

  * Support for storing Musicbrainz metadata in the multimedia ontology.
  * Doc cleanups
  * Test suite improvements
  * Build fixes
  * Fix detection of files that need writeback
  * Fix crashes and invalid memory writes

NEW in 2.2.2 - 2019-05-02
=========================

  * Fixed initialization of virtual tables
  * Fixed segmentation fault in libtracker-miner
  * Don't try to create JSON-LD nodes with unsigned integers
  * Handle correctly backreferences in TrackerResource tree
  * Many documentation fixes
  * Fixed handling doubles with exponents in SPARQL
  * Don't limit to specific desktop environments
  * Build and test fixes

Translations: hr

NEW in 2.2.1 - 2019-03-06
=========================

  * Fix unichar unescaping
  * Correctly Handle BIND in first place of a triples block
  * Fix possible deadlock on WAL checkpoint
  * Fix some double values not being deleted
  * Fixed CHANGES_DONE_HINT handling in TrackerMonitor
  * Ported data generator utilities to python3
  * Ported functional tests to python3, reformatted to PEP-8
  * Correctly apply ignored-directories-with-content filter on monitor updates
  * Build and test fixes

Translations: de, eo, fur, fr, it, ko, lv, nl, ru, sr

NEW in 2.2.0 - 2019-02-20
=========================

  * Multiple memory leak and corruption fixes
  * Bumped glib minimum version to 2.46, it already was in practical terms
  * Test suite improvements
  * Restore log domain

Translations: ca, cs, da, kk, lt, pt_BR, sv, zh_TW

  Highlighted changes since 2.1.x:
  --------------------------------

  * New SPARQL parser, able to generate SQL that is generally more readable
    and at places performs better. Multiple buglets fixed in the process
  * Much improved support of SPARQL1.1 features and syntax that was missing:
    - Property paths: Allowing to match connectivity between two resources
      by an arbitrary length path. There is a number of supported operators
      (alternative, sequence, oneOrMany, ...) that can be combined, e.g:

      SELECT ?s ?p { ?s ^(nfo:belongsToContainer*)/(nie:url|nie:title) ?p }

      Only the negated path operator (!) is not supported at the moment.

    - Support for fully unrestricted queries, eg:

      SELECT ?s ?p ?o { ?s ?p ?o } ORDER BY ?o ?p ?s

      Queries with unrestricted predicate (?p in the example above) were
      just supported in a very restricted set of situations. All those
      limitations are gone.
    - MINUS allows subtracting the solutions that match the given triples
      template, eg:

      SELECT ?s { ?s a nfo:Media } MINUS { ?s a nfo:MusicPiece }

  * Support for prepared statements. TrackerSparqlStatement can be built
    with SELECT queries containing (custom) ~var syntax, and updating
    their values before obtaining a cursor.
  * Many tests were added, and Tracker is generally much better tested thanks
    to CI.
  * tracker-store now automatically shuts down on inactivity.

  NOTE TO PACKAGERS: This release might trigger ABI checks, as
    TrackerSparqlConnectionClass struct size grew. This object can only
    be subclassed within Tracker and instantiated through Tracker API, so
    the change is inconsequential.

NEW in 2.2.0-alpha2 - 2019-02-06
================================

  * More property paths supported, new operators supported are *, +, ? and |,
    only the ! operator is not supported yet.
  * Multiple fixes to the new sparql parser.
  * Improve error handling in DBus backend
  * Many improvements and additions to the test suite
  * Fixed potential crash in TrackerMonitor.

Translations: el, es, fur, fr, gl, id, pl, ro, sl, sv, tr

NEW in 2.2.0-alpha1 - 2018-11-13
================================

  * New SPARQL parser, able to support more 1.1 features and generating
    friendlier SQL at places. There is initial support for property
    paths (/ and ^), and other missing 1.1 syntax (MINUS, SHA384, ...).
    More improvements are expected to happen in the future thanks to this.
  * Support for prepared statements. TrackerSparqlStatement can be built
    with SELECT queries containing (custom) ~var syntax, and updating
    their values before obtaining a cursor.
  * Added global libtracker-sparql call to change the used DBus connection
    at runtime.
  * Made tracker-store to automatically shutdown when unneeded.

src/tracker/tracker-export.c
/*
 * Copyright (C) 2020, Sam Thursfield <sam@afuera.me.uk>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 */

#include "config.h"

#include <sys/param.h>
#include <stdlib.h>
#include <time.h>
#include <locale.h>

#include <glib.h>
#include <glib/gi18n.h>

#include <libtracker-sparql/tracker-sparql.h>

#include "tracker-sparql.h"
#include "tracker-color.h"

static gboolean show_graphs;

static GOptionEntry entries[] = {
	{ "show-graphs", 'g', 0, G_OPTION_ARG_NONE, &show_graphs,
	  N_("Output TriG format which includes named graph information"),
	  NULL
	},
	{ NULL }
};

static TrackerSparqlConnection *
create_connection (GError **error)
{
	return tracker_sparql_connection_get (NULL, error);
}

/* format a URI for Turtle; if it has a prefix, display uri
 * as prefix:rest_of_uri; if not, display as <uri>
 */
inline static gchar *
format_urn (GHashTable  *prefixes,
            const gchar *urn,
            gboolean     full_namespaces)
{
	gchar *urn_out;

	if (full_namespaces) {
		urn_out = g_strdup_printf ("<%s>", urn);
	} else {
		gchar *shorthand = tracker_sparql_get_shorthand (prefixes, urn);

		/* If the shorthand is the same as the urn passed, we
		 * assume it is a resource and pass it in as one,
		 *
		 *   e.g.: http://purl.org/dc/elements/1.1/date
		 *     to: http://purl.org/dc/elements/1.1/date
		 *
		 * Otherwise, we use the shorthand version instead.
		 *
		 *   e.g.: http://www.w3.org/1999/02/22-rdf-syntax-ns
		 *     to: rdf
		 */
		if (g_strcmp0 (shorthand, urn) == 0) {
			urn_out = g_strdup_printf ("<%s>", urn);
			g_free (shorthand);
		} else {
			urn_out = shorthand;
		}
	}

	return urn_out;
}

/* print a URI prefix in Turtle format */
static void
print_prefix (gpointer key,
              gpointer value,
              gpointer user_data)
{
	g_print ("@prefix %s: <%s#> .\n", (gchar *) value, (gchar *) key);
}

/* Print triples in Turtle format */
static void
print_turtle (TrackerSparqlCursor *cursor,
              GHashTable          *prefixes,
              gboolean             full_namespaces)
{
	gchar *predicate;
	gchar *object;

	while (tracker_sparql_cursor_next (cursor, NULL, NULL)) {
		const gchar *resource = tracker_sparql_cursor_get_string (cursor, 1, NULL);
		const gchar *key = tracker_sparql_cursor_get_string (cursor, 2, NULL);
		const gchar *value = tracker_sparql_cursor_get_string (cursor, 3, NULL);
		const gchar *value_is_resource = tracker_sparql_cursor_get_string (cursor, 4, NULL);

		if (!resource || !key || !value || !value_is_resource) {
			continue;
		}

		/* Don't display nie:plainTextContent */
		//if (!plain_text_content && strcmp (key, "http://www.semanticdesktop.org/ontologies/2007/01/19/nie#plainTextContent") == 0) {
		//	continue;
		//}

		predicate = format_urn (prefixes, key, full_namespaces);

		if (g_ascii_strcasecmp (value_is_resource, "true") == 0) {
			object = g_strdup_printf ("<%s>", value);
		} else {
			gchar *escaped_value;

			/* Escape value and make sure it is encapsulated properly */
			escaped_value = tracker_sparql_escape_string (value);
			object = g_strdup_printf ("\"%s\"", escaped_value);
			g_free (escaped_value);
		}

		/* Print final statement */
		g_print ("<%s> %s %s .\n", resource, predicate, object);

		g_free (predicate);
		g_free (object);
	}
}

/* Print graphs and triples in TriG format */
static void
print_trig (TrackerSparqlCursor *cursor,
            GHashTable          *prefixes,
            gboolean             full_namespaces)
{
	gchar *predicate;
	gchar *object;
	gchar *previous_graph = NULL;
	const gchar *graph;

	while (tracker_sparql_cursor_next (cursor, NULL, NULL)) {
		graph = tracker_sparql_cursor_get_string (cursor, 0, NULL);
		const gchar *resource = tracker_sparql_cursor_get_string (cursor, 1, NULL);
		const gchar *key = tracker_sparql_cursor_get_string (cursor, 2, NULL);
		const gchar *value = tracker_sparql_cursor_get_string (cursor, 3, NULL);
		const gchar *value_is_resource = tracker_sparql_cursor_get_string (cursor, 4, NULL);

		if (!resource || !key || !value || !value_is_resource) {
			continue;
		}

		if (g_strcmp0 (previous_graph, graph) != 0) {
			if (previous_graph != NULL) {
				/* Close previous graph */
				g_print ("}\n");
				g_free (previous_graph);
			}
			previous_graph = g_strdup (graph);
			g_print ("GRAPH <%s>\n{\n", graph);
		}

		/* Don't display nie:plainTextContent */
		//if (!plain_text_content && strcmp (key, "http://tracker.api.gnome.org/ontology/v3/nie#plainTextContent") == 0) {
		//	continue;
		//}

		predicate = format_urn (prefixes, key, full_namespaces);

		if (g_ascii_strcasecmp (value_is_resource, "true") == 0) {
			object = g_strdup_printf ("<%s>", value);
		} else {
			gchar *escaped_value;

			/* Escape value and make sure it is encapsulated properly */
			escaped_value = tracker_sparql_escape_string (value);
			object = g_strdup_printf ("\"%s\"", escaped_value);
			g_free (escaped_value);
		}

		/* Print final statement */
		g_print ("  <%s> %s %s .\n", resource, predicate, object);

		g_free (predicate);
		g_free (object);
	}

	if (graph != NULL) {
		g_print ("}\n");
	}
	g_free (previous_graph);
}

static int
export_run_default (void)
{
	g_autoptr(TrackerSparqlConnection) connection = NULL;
	g_autoptr(TrackerSparqlCursor) cursor = NULL;
	g_autoptr(GError) error = NULL;
	g_autoptr(GHashTable) prefixes = NULL;
	const gchar *query;

	connection = create_connection (&error);

	if (!connection) {
		g_printerr ("%s: %s\n",
		            _("Could not establish a connection to Tracker"),
		            error ? error->message : _("No error given"));
		return EXIT_FAILURE;
	}

	prefixes = tracker_sparql_get_prefixes ();

	query = "SELECT ?g ?u ?p ?v "
	        "       (EXISTS { ?p rdfs:range [ rdfs:subClassOf rdfs:Resource ] }) AS ?is_resource "
	        "{ "
	        "    GRAPH ?g { "
	        "        ?u ?p ?v "
	        "        FILTER NOT EXISTS { ?u a rdf:Property } "
	        "        FILTER NOT EXISTS { ?u a rdfs:Class } "
	        "        FILTER NOT EXISTS { ?u a tracker:Namespace } "
	        "    } "
	        "} ORDER BY ?g ?u";

	cursor = tracker_sparql_connection_query (connection, query, NULL, &error);

	if (error) {
		g_printerr ("%s, %s\n",
		            _("Could not run query"),
		            error->message);
		return EXIT_FAILURE;
	}

	g_hash_table_foreach (prefixes, (GHFunc) print_prefix, NULL);
	g_print ("\n");

	if (show_graphs) {
		print_trig (cursor, prefixes, FALSE);
	} else {
		print_turtle (cursor, prefixes, FALSE);
	}

	return EXIT_SUCCESS;
}

int
tracker_export (int argc, const char **argv)
{
	GOptionContext *context;
	GError *error = NULL;

	context = g_option_context_new (NULL);
	g_option_context_add_main_entries (context, entries, NULL);

	argv[0] = "tracker export";

	if (!g_option_context_parse (context, &argc, (char***) &argv, &error)) {
		g_printerr ("%s, %s\n", _("Unrecognized options"), error->message);
		g_error_free (error);
		g_option_context_free (context);
		return EXIT_FAILURE;
	}

	g_option_context_free (context);

	return export_run_default ();
}
